src/centralized_telescope.rs

```diff
@@ -0,0 +1,419 @@
+    fn h1(
+        first_input: &[u8],
+        second_input: &[u8],
+        n_p: u64,
+        sec_param: u64,
+    ) -> (Hash, Option<u64>) {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H1");
+        hasher.update(first_input);
+        hasher.update(second_input);
+        let digest: Hash = hasher.finalize().into();
+        (digest, utils::sample_uniform(&digest, n_p, sec_param))
+    }
+
+    /// Output a round from a proof counter and n_p
+    /// Initilialises the hash with H1(t) and random value as oracle(H1(t), n_p)
+    pub fn new(v: u64, t: u64, n_p: u64, sec_param: u64) -> Option<Self> {
+        let (h, h_u64_opt) = Self::h1(
+            v.to_be_bytes().as_ref(),
+            t.to_be_bytes().as_ref(),
+            n_p,
+            sec_param,
+        );
+        h_u64_opt.map(|h_u64| Self {
+            v,
+            t,
+            s_list: vec![],
+            h,
+            h_u64,
+            n_p,
+        })
+    }
+
+    /// Updates a round with an element of S_p
+    /// Replaces the hash $h$ with $h' = H1(h, s)$ and the random value as oracle(h', n_p)
+    pub fn update(r: &Self, s: Element, sec_param: u64) -> Option<Self> {
+        let mut s_list = r.s_list.clone();
+        s_list.push(s);
+        let (h, h_u64_opt) = Self::h1(r.h.clone().as_ref(), s.as_ref(), r.n_p, sec_param);
+        h_u64_opt.map(|h_u64| Self {
+            v: r.v,
+            t: r.t,
+            s_list,
+            h,
+            h_u64,
+            n_p: r.n_p,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+/// Alba proof
+pub struct Proof {
+    /// Proof counter
+    v: u64,
+    /// Proof 2nd counter
+    t: u64,
+    /// Proof tuple
+    items: Vec<Element>,
+}
+
+impl Proof {
+    /// Oracle producing a uniformly random value in [1, n_p] used for prehashing S_p
+    fn h0(setup: &Setup, v: u64, s: Element) -> Option<u64> {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H0");
+        hasher.update(v.to_be_bytes());
+        hasher.update(s);
+        let digest: Hash = hasher.finalize().into();
+        utils::sample_uniform(&digest, setup.n_p, setup.sec_param)
+    }
+
+    /// Oracle defined as Bernoulli(q) returning 1 with probability q and 0 otherwise
+    fn h2(setup: &Setup, r: &Round) -> bool {
+        let mut hasher = Blake2s256::new();
+        hasher.update(b"Telescope-H2");
+        hasher.update(r.h);
+        let digest: Hash = hasher.finalize().into();
+        utils::sample_bernouilli(&digest, setup.q, setup.sec_param)
+    }
+
+    /// Depth-first search which goes through all potential round candidates
+    /// and returns first round candidate Round{t, x_1, ..., x_u)} such that:
+    /// - for all i ∈ [0, u-1], H0(x_i+1) ∈ bins[H1(t, x_1, ..., x_i)]
+    /// - H2(t, x_0, ..., x_u) = true
+    fn dfs(setup: &Setup, bins: &[Vec<Element>], round: &Round, limit: u64) -> (u64, Option<Self>) {
```

Comment:
!Future!
I remember that I commented that we should comply with the names in the paper.
However, `limit` does not seem to cover its actual functionality.

Reply:
Yes, you wanted to change the name from "steps" to "limit". I had argued that here we are incrementing the value and not decreasing it contrary to the paper. Perhaps we can add this in the variable name issue?